DOC Improve `SGDClassifier` docstrings

poorna-kumar opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Related to https://github.com/scikit-learn/scikit-learn/issues/15440. 

#### What does this implement/fix? Explain your changes.

Adds `numpydoc` validation to some of `SGDClassifier`'s docstrings. 

#### Any other comments?

Couldn't fix make `SGDClassifier` fully compliant with `numpydoc`; ran into linter issues with `SGDClassifier.loss_functions`, `SGDClassifier.predict_log_proba` and `SGDClassifier.predict_proba`. (Possibly because none of them are "regular" methods?)

Example issue for `SGDClassifier.loss_functions` below:

```
> python maint_tools/test_docstrings.py sklearn.linear_model.SGDClassifier.loss_functions
Traceback (most recent call last):
  File "maint_tools/test_docstrings.py", line 173, in <module>
    msg = repr_errors(res, method=args.import_path)
  File "maint_tools/test_docstrings.py", line 112, in repr_errors
    for code, message in res["errors"]
TypeError: sequence item 0: expected str instance, NoneType found
```
<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

poorna-kumar committed:
fix fit method docstring

poorna-kumar committed:
fix partial_fit method docstring

poorna-kumar committed:
add docstring to SGDClassifier.set_params()

TomDLT commented:
Thanks for this pull-request !

Please mention the issue your are addressing in the pull-request message, in this case #15440 I guess.

If you finished fixing the doc of one estimator, can you add the estimator to the whitelist in `maint_tools/test_docstrings.py` [here](https://github.com/scikit-learn/scikit-learn/blob/master/maint_tools/test_docstrings.py#L11)?

poorna-kumar committed:
fix docstring for SGDClassifier class

poorna-kumar committed:
remove docstrings for params that are not passed

poorna-kumar committed:
flake8: fix line length

poorna-kumar committed:
add functions with fixed docstrings to whitelist

amueller commented:
this has a merge conflict now (because your other PR was merged).

poorna-kumar committed:
make a change to trigger test run again

amueller commented:
Is removing the parameter X from predict_proba actually helpful? I'm a bit skeptical.

poorna-kumar commented:
You're right that it's not ideal to remove the parameter X -- let me try doing something else to appease the linter, and if that doesn't work, will remove this method from the whitelist.

poorna-kumar committed:
Undo changes to documentation for SGDClassifier.predict_log_proba and SGDClassifier.predict_proba
1) While the previous changes were technically correct, they could have misled readers.
2) Provide a clarifying note to explain what the function is doing.

poorna-kumar committed:
remove predict_proba and predict_log_proba from whitelist

poorna-kumar committed:
add clarifying comment

rth commented on sklearn/linear_model/_stochastic_gradient.py:
```diff
@@ -1032,9 +1058,17 @@ def predict_log_proba(self):
+        ``_predict_log_proba``. However, since it returns a callable
+        method, when it's invoked in code, it appears
+        like it's being called with an argument (of input data) X, and
+        returning probabilities T.
```
That is true, but it's an implementation detail.  I thin what is said at the top,

> This method is only available for log loss and modified Huber loss.

Is enough from the user perspective, so I would remove this note.

rth commented on sklearn/linear_model/_stochastic_gradient.py:
```diff
@@ -1032,9 +1058,17 @@ def predict_log_proba(self):
         Parameters
         ----------
-        X : array-like, shape (n_samples, n_features)
+        X : {array-like, sparse matrix}, shape (n_samples, n_features)
```
```suggestion
        X : {array-like, sparse matrix} of shape (n_samples, n_features)
```

poorna-kumar replied:
Made this change, although it's inconsistent with other similar docstrings in the file (like [this](https://github.com/scikit-learn/scikit-learn/blob/968252dfb4c252e252614ae8736cc6b1ebdde9ee/sklearn/linear_model/_stochastic_gradient.py#L1209)).

rth commented on sklearn/linear_model/_stochastic_gradient.py:
```diff
@@ -955,9 +973,17 @@ def predict_proba(self):
+        ``_predict_proba``. However, since it returns a callable
+        method, when it's invoked in code, it appears
+        like it's being called with an argument (of input data) X, and
+        returning probabilities T.
```
Same as the other comment.

rth commented on sklearn/linear_model/_stochastic_gradient.py:
```diff
@@ -99,6 +99,19 @@ def __init__(self, loss, penalty='l2', alpha=0.0001, C=1.0,
 
     def set_params(self, **kwargs):
+        """
+        Set and validate the parameters of estimator.
```
This should be,
```py
"""Set and validate
```
we disabled the rule numpy that would have produced this output.

rth reviewed (changes requested):
Thanks @poorna-kumar ! A few comments below.

poorna-kumar committed:
reviewer comments: remove side note

poorna-kumar committed:
reviewer comments: add double backticks around SGDClassifier

cmarmo commented:
@poorna-kumar do you have some time to synchronize with upstream and address comments?
Thanks!

poorna-kumar committed:
reviewer's comments on docstring

cmarmo commented:
Hi @rth ,let us know if you are happy with that... :)

rth reviewed (approved):
Thanks @poorna-kumar and @cmarmo !
